### PR TITLE
Add custom Newton's method example

### PR DESCRIPTION
**Pull request by jorgensd**
Addresses #60. @bhaveshshrimali could you have a look and see what you think?

### CODE REVIEW

File: chapter4/newton-solver.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    solver.solve(L, dx.vector)
+    dx.x.scatter_forward()
+    # Update u_{i+1} = u_i + delta x_i
+    uh.x.array[:] += dx.x.array
+    i += 1
+
+    # Compute norm of update
+    correction_norm = dx.vector.norm(0)
+    print(f"Iteration {i}: Correction norm {correction_norm}")
+    if correction_norm < 1e-10:
```
[the remark this answers is not in the document]
Comment (bhaveshshrimali) [reply chapter4/newton-solver.py:128]: I didn't see that it is verified further down the line that `L.norm(0)` is indeed small. Either way it fine!

File: chapter4/newton-solver.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+    # Compute norm of update
+    correction_norm = dx.vector.norm(0)
+
+    # Compute L2 error comparing to the analytical solution
+    L2_error.append(np.sqrt(mesh.comm.allreduce(dolfinx.fem.assemble_scalar(error), op=MPI.SUM)))
+    dx_norm.append(correction_norm)
+    
+    print(f"Iteration {i}: Correction norm {correction_norm}, L2 error: {L2_error[-1]}")
+    if correction_norm < 1e-10:
```
Comment (bhaveshshrimali) [inline chapter4/newton-solver.py:260]: I am guessing this corresponds to "incremental" as the `convergence_criterion` when using an inbuilt newton solver. 

[Update] Indeed it is!

**Review by bhaveshshrimali — COMMENTED** [review]
Looks perfect! Thanks for putting it together @jorgensd 
 
 It would be a good resource for anyone trying to implement their own Newton's method/nonlinear solver

File: chapter4/newton-solver.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jorgensd) [reply chapter4/newton-solver.py:128]: There are many ways of terminering a Newton solver. I might add a section on that later, as one could Also compute the residual of the Newton solver (Which is different to L(uh), if uh does not satisfy the boundary condition.

**Pull request merged**